src/firejail/template.c

```diff
@@ -0,0 +1,429 @@
+
+	template_free(tmpl);
+
+out:
+	free(macro);
+
+	if (err)
+		exit(-err);
+}
+
+static char* append_to_string(char *str1, const char *str2)
+{
+	size_t len;
+
+	if (!str2)
+		return str1;
+
+	if (!str1)
+		return strdup(str2);
+
+	len = strlen(str2) + 1;
+	str1 = (char*)realloc(str1, strlen(str1) + len);
+
+	return strncat(str1, str2, len);
+}
+
+const char *internal_keys[] = { "HOME", "CFG", "RUNUSER", "PATH", "PRIVILEGED",
+				NULL };
+
+static int is_internal_macro(const char *key)
+{
+	char *macro;
+	char *macro_value;
+	int err = 0;
+	int i;
+
+	for (i = 0; internal_keys[i]; i++) {
+		if (!strncmp(key, internal_keys[i], strlen(internal_keys[i]))) {
+			err = -EEXIST;
+			goto exit;
+		}
+	}
+
+	if (asprintf(&macro, "${%s}", key) == -1)
+		errExit("asprintf");
+
+	macro_value = resolve_macro(key);
+	if (macro_value) {
+		err = -EEXIST;
+		free(macro_value);
+	}
+
+	free(macro);
+exit:
+	return err;
+}
+
+static char *process_key_value(char *container, char *key, int *err)
+{
+	char *str;
+	char *token;
+	char *saveptr;
+	const char *delim = TEMPLATE_KEY_MACRO_SUB_DELIMS;
+	const char *value;
+
+	for (str = key; ; str = NULL) {
+		token = strtok_r(str, delim, &saveptr);
+		if (!token)
+			break;
+
+		if (is_internal_macro(token) == -EEXIST) {
+			char *macro;
+
+			if (asprintf(&macro, "${%s}", token) == -1)
+				errExit("asprintf");
+
+			container = append_to_string(container, macro);
+			free(macro);
+			continue;
+		}
+
+		/* Only the first iteration is the template */
+		value = str ? template_get_value(token, err) : token;
+		if (!value)
+			goto err;
+
+		container = append_to_string(container, value);
+		if (!container)
+			goto err;
+	}
+
+	return container;
+
+err:
+	if (container)
+		free(container);
+	else
+		*err = -EINVAL;
+
+	return NULL;
+}
+
+/*
+ * Allocates new string with all template keus replaced with the values.
+ * Returns NULL if there is a nonexistent key, allocation fails or if arg
+ * begins with $. If arg does not contain $ it is only duplicated.
+ */
+char *template_replace_keys(char *arg, int *err)
+{
+	char *new_string = NULL;
+	char *str;
+	char *token;
+	char *saveptr;
+	const char *delim = TEMPLATE_KEY_MACRO_DELIM;
+
+	if (!arg || !*arg) {
+		*err = -ENOENT;
+		return NULL;
+	}
+
+	if (!strrchr(arg, '$')) {
+		fprintf(stderr, "No \"$\" in \"%s\" return copy\n", arg);
+		return strdup(arg);
+	}
+
+	/* Templates must not be given at the beginning of the argument */
+	if (*arg == '$') {
+		*err = -EINVAL;
+		fprintf(stderr, "Starts with \"$\" in \"%s\"\n", arg);
+		return NULL;
+	}
+
+	for (str = arg; ; str = NULL) {
+		token = strtok_r(str, delim, &saveptr);
+		if (!token)
+			break;
+
+		/*
+		 * Process template values starting from the second token as
+		 * templates cannot be used at the beginning of the arg
+		 * because only hardcoded macros should be as first.
+		 */
+		if (!str)
+			new_string = process_key_value(new_string, token, err);
+		else
+			new_string = append_to_string(new_string, token);
```

Comment:
Will add check that if there is a lone `$` it means an error. If that is any of the internal macros it will be replaced in the string, yeah, not the best way but since I resorted using `strtok_r()` I guess that is unavoidable unless that is to be done by hand, which may induce more errors..

See also https://github.com/LaakkonenJussi/firejail/pull/1/files#r630917495

Reply:
I'd suppose this is resolved but @spiiroin  do you still have something that needs to be clarified or do you think this needs fixing? I added comments to code and to commit msg about this as well.